File: KeySAVCoreJS/scripts/source/SaveReaderEncrypted.cs
using System;
using System.IO;
using KeySAVCore.Structures;
using JS;

namespace KeySAVCore
{
    public class SaveReaderEncrypted : ISaveReader
    {
        private readonly Uint8Array sav;
        private readonly SaveKey key;
        private readonly byte activeSlot;
        private readonly string _KeyName;

        public string KeyName
        {
            get { return _KeyName; }
        }

        public ushort UnlockedSlots
        {
            get
            {
                ushort res = 0;
                for (int i = 0; i < key.slotsUnlocked.length; ++i)
                    if (key.slotsUnlocked[i]!= 0)
                        ++res;
                return res;
            }
        }

        // TODO unlocked slots kinda broken

        private readonly static Uint8Array zeros;
        private readonly static Uint8Array ezeros;

        internal SaveReaderEncrypted(Uint8Array file, Uint8Array key_)
        {
            sav = file;

            key = new SaveKey(key_);

            _KeyName = Path.GetFileName(_KeyName);

            if (key.slot1Flag == BitConverter.ToUInt32(sav, 0x168))
                activeSlot = 0;
            else
                activeSlot = 1;

            sav.XorInPlace((int)key.boxOffset-0x7F000, key.slot1Key, 0, 232*30*31);
        }

        internal SaveReaderEncrypted(Uint8Array file, SaveKey key)
        {
            sav = file;

            this.key = key;

            _KeyName = "";

            if (key.slot1Flag == BitConverter.ToUInt32(sav, 0x168))
                activeSlot = 0;
            else
                activeSlot = 1;

            sav.XorInPlace((int)key.boxOffset-0x7F000, key.slot1Key, 0, 232*30*31);
        }
        
        static SaveReaderEncrypted()
        {
            zeros = new Uint8Array(232);
            ezeros = PKX.encrypt(zeros);
        }

        public void scanSlots()
        {
            bool ghost;
            for(ushort i = 0; i < 30*31; ++i)
            {
                getPkxRaw(i, 0, out ghost);
                getPkxRaw(i, 1, out ghost);
            }
        }

        public void scanSlots(ushort pos)
        {
            bool ghost;
            getPkxRaw(pos, 0, out ghost);
            getPkxRaw(pos, 1, out ghost);
        }

        public void scanSlots(ushort start, ushort end)
        {
            for (ushort i = start; i <= end; ++i)
            {
                bool ghost;
                getPkxRaw(i, 0, out ghost);
                getPkxRaw(i, 1, out ghost);
            }
        }

        public PKX? getPkx(ushort pos)
        {
            bool ghost;
            Uint8Array data = getPkxRaw(pos, activeSlot, out ghost);
            if (data == null || (data[8] | data[9]) == 0)
                return null;
            return new PKX(data, (byte)(pos/30), (byte)(pos%30), ghost);
        }

        // Original code by Kaphotics
        private Uint8Array getPkxRaw(ushort pos, byte slot, out bool ghost)
        {
            // TODO refactor moar
            // Auto updates the keystream when it dumps important data!
            Uint8Array ekx = new Uint8Array(232);

            ghost = true;

            int keyOffset = pos * 232;
            int savOffset = (int)(keyOffset + key.boxOffset - (1 - slot) * 0x7F000);

            if (zeros.SequenceEqual(key.boxKey1, keyOffset) && zeros.SequenceEqual(key.boxKey2, keyOffset))
                return null;
            else if (zeros.SequenceEqual(key.boxKey1, keyOffset))
            {
                // Key2 is confirmed to dump the data.
                ekx = Utility.xor(key.boxKey2, keyOffset, sav, savOffset, 232);
                ghost = false;
                key.slotsUnlocked[slot] = 1;
            }
            else if (zeros.SequenceEqual(key.boxKey2, keyOffset))
            {
                // Haven't dumped from this slot yet.
                if (Utility.SequenceEqual(key.boxKey1, keyOffset, sav, savOffset, 232))
                {
                    // Slot hasn't changed.
                    return null;
                }
                else
                {
                    // Try and decrypt the data...
                    ekx = Utility.xor(key.boxKey1, keyOffset, sav, savOffset, 232);
                    if (PKX.verifyCHK(PKX.decrypt(ekx)))
                    {
                        // Data has been dumped!
                        // Fill keystream data with our log.
                        Uint8ArrayHelper.Copy(sav, savOffset, key.boxKey2, keyOffset, 232);
                    }
                    else
                    {
                        // Try xoring with the empty data.
                        if (PKX.verifyCHK(PKX.decrypt(Utility.xor(ekx, key.blank))))
                        {
                            ekx = Utility.xor(ekx, key.blank);
                            Utility.xor(key.blank, 0, sav, savOffset, key.boxKey2, keyOffset, 232);
                        }
                        else if (PKX.verifyCHK(PKX.decrypt(Utility.xor(ekx, ezeros))))
                        {
                            ekx = Utility.xor(ekx, ezeros);
                            Utility.xor(ezeros, 0, sav, savOffset, key.boxKey2, keyOffset, 232);
                        }
                        else return null; // Not a failed decryption; we just haven't seen new data here yet.
                    }
                }
            }
            else
            {
                // We've dumped data at least once.
                if (Utility.SequenceEqual(key.boxKey1, keyOffset, sav, savOffset, 232) ||
                    Utility.SequenceEqual(key.boxKey1, keyOffset, Utility.xor(key.blank, sav, savOffset), 0, 232) ||
                    Utility.SequenceEqual(key.boxKey1, keyOffset, Utility.xor(ezeros, sav, savOffset), 0, 232))
                {
                    // Data is back to break state, but we can still dump with the other key.
                    ekx = Utility.xor(key.boxKey2, keyOffset, sav, savOffset, 232);
                    if (!PKX.verifyCHK(PKX.decrypt(ekx)))
                    {
                        if (PKX.verifyCHK(PKX.decrypt(Utility.xor(ekx, key.blank))))
                        {
                            ekx = Utility.xor(ekx, key.blank);
                            Utility.xor(key.blank, 0, key.boxKey2, keyOffset, key.boxKey2, keyOffset, 232);
                        }
                        else if (PKX.verifyCHK(PKX.decrypt(Utility.xor(ekx, ezeros))))
                        {
                            // Key1 decrypts our data after we remove encrypted zeros.
                            // Copy Key1 to Key2, then zero out Key1.
                            ekx = Utility.xor(ekx, ezeros);
                            Utility.xor(ezeros, 0, key.boxKey2, keyOffset, key.boxKey2, keyOffset, 232);
                        }
                        else return null; // Decryption Error
                    }
                }
                else if (Utility.SequenceEqual(key.boxKey2, keyOffset, sav, savOffset, 232) ||
                    Utility.SequenceEqual(key.boxKey2, keyOffset, Utility.xor(key.blank, sav, savOffset), 0, 232) ||
                    Utility.SequenceEqual(key.boxKey2, keyOffset, Utility.xor(ezeros, sav, savOffset), 0, 232))
                {
                    // Data is changed only once to a dumpable, but we can still dump with the other key.
                    ekx = Utility.xor(key.boxKey1, keyOffset, sav, savOffset, 232); 
                    if (!PKX.verifyCHK(PKX.decrypt(ekx)))
                    {
                        if (PKX.verifyCHK(PKX.decrypt(Utility.xor(ekx, key.blank))))
                        {
                            ekx = Utility.xor(ekx, key.blank);
                            Utility.xor(key.blank, 0, key.boxKey1, keyOffset, key.boxKey1, keyOffset, 232);
                        }
                        else if (PKX.verifyCHK(PKX.decrypt(Utility.xor(ekx, ezeros))))
                        {
                            ekx = Utility.xor(ekx, ezeros);
                            Utility.xor(ezeros, 0, key.boxKey1, keyOffset, key.boxKey1, keyOffset, 232);
                        }
                        else return null; // Decryption Error
                    }
                }
                else
                {
                    // Data has been observed to change twice! We can get our exact keystream now!
                    // Either Key1 or Key2 or Save is empty. Whichever one decrypts properly is the empty data.
                    // Oh boy... here we go...
                    ghost = false;
                    key.slotsUnlocked[slot] = 1;
                    bool keydata1, keydata2 = false;
                    Uint8Array data1 = Utility.xor(sav, savOffset, key.boxKey1, keyOffset, 232);
                    Uint8Array data2 = Utility.xor(sav, savOffset, key.boxKey2, keyOffset, 232);

                    keydata1 = 
                        (PKX.verifyCHK(PKX.decrypt(data1))
                        ||
                        PKX.verifyCHK(PKX.decrypt(Utility.xor(data1, ezeros)))
                        ||
                        PKX.verifyCHK(PKX.decrypt(Utility.xor(data1, key.blank)))
                        );
                    keydata2 = 
                        (PKX.verifyCHK(PKX.decrypt(data2))
                        ||
                        PKX.verifyCHK(PKX.decrypt(Utility.xor(data2, ezeros)))
                        ||
                        PKX.verifyCHK(PKX.decrypt(Utility.xor(data2, key.blank)))
                        );

                    Uint8Array emptyKey, emptyKeyData;
                    int emptyOffset;

                    if (keydata1 && keydata2)
                    {
                        // Save file is currently empty...
                        // Copy key data from save file if it decrypts with Key1 data properly.
                        emptyKey = sav;
                        emptyOffset = savOffset;
                        emptyKeyData = data1;
                    }
                    else if (keydata1) // Key 1 is empty
                    {
                        emptyKey = key.boxKey1;
                        emptyOffset = keyOffset;
                        emptyKeyData = data1;
                    }
                    else if (keydata2) // Key 2 is emtpy
                    {
                        emptyKey = key.boxKey2;
                        emptyOffset = keyOffset;
                        emptyKeyData = data2;
                    }
                    else return null; // All three are occupied

                    if (PKX.verifyCHK(PKX.decrypt(emptyKeyData)))
                    {
                        // No modifications necessary.
                        ekx = emptyKeyData;
                        Uint8ArrayHelper.Copy(emptyKey, emptyOffset, key.boxKey2, keyOffset, 232);
                        Uint8ArrayHelper.Copy(zeros, 0, key.boxKey1, keyOffset, 232);
                    }
                    else if (PKX.verifyCHK(PKX.decrypt(Utility.xor(emptyKeyData, ezeros))))
                    {
                        ekx = ezeros;
                        Utility.xor(ezeros, 0, emptyKey, emptyOffset, key.boxKey2, keyOffset, 232);
                        Uint8ArrayHelper.Copy(zeros, 0, key.boxKey1, keyOffset, 232);
                    }
                    else if (PKX.verifyCHK(PKX.decrypt(Utility.xor(emptyKeyData, key.blank))))
                    {
                        ekx = ezeros;
                        Utility.xor(key.blank, 0, emptyKey, emptyOffset, key.boxKey2, keyOffset, 232);
                        Uint8ArrayHelper.Copy(zeros, 0, key.boxKey1, keyOffset, 232);
                    }
                }
            }
            Uint8Array pkx = PKX.decrypt(ekx);
            if (PKX.verifyCHK(pkx))
            {
                return pkx;
            }
            else 
                return null; // Slot Decryption error?!

        }
    }
}

File: save-reader-encrypted.ts
import SaveReader from "./save-reader";
import SaveKey from "./save-key";
import PkBase from "./pkbase";
import * as util from "./util";

var zeros = new Uint8Array(232);
var ezeros = PkBase.encrypt(zeros);

export { default as SaveReader } from "./save-reader";
export default class SaveReaderEncrypted implements SaveReader {
    public static getOffsets(generation: number) {
        return {
            6: {
                fileSize: 0x100000,
                saveSize: 0x07f000,
                base1: 0x1000,
                base2: 0x080000
            },
            7: {
                fileSize: 0x0fe000,
                saveSize: 0x07e000,
                base1: 0x2000,
                base2: 0x080000
            }
        }[generation];
    }

    public static getGeneration(file: Uint8Array) {
        let length = file.length;
        if (length === 0x100000 || length === 0x10009C || length === 0x10019A) {
            return 6;
        }
        if (length === 0x0fe000 || length === 0x0fe09c || length === 0x0fe19a) {
            return 7;
        }
        return -1;
    }

    private activeSlot: number;
    private boxes1: Uint8Array;
    private boxes2: Uint8Array;

    public get generation() {
        return this.key.generation;
    }

    get unlockedSlots() {
        var res = 0;
        for (var i = 0; i < this.key.slotsUnlocked.length; ++i)
            if (this.key.slotsUnlocked[i])
                ++res;
        return res;
    }

    get isNewKey() {
        return this.key.isNewKey;
    }

    constructor(private sav: Uint8Array, private key: SaveKey) {
        const offsets = SaveReaderEncrypted.getOffsets(this.generation);

        this.sav = this.sav.subarray(this.sav.length % offsets.fileSize);
        this.activeSlot = this.key.slot1Flag == util.createDataView(sav).getUint32(0x168, true) && this.key.isNewKey ? 0 : 1;
        this.boxes1 = util.xor(this.sav, key.boxOffset - offsets.saveSize, key.slot1Key, 0, 232 * 30 * (this.generation === 6 ? 31 : 32));
        this.boxes2 = this.sav.subarray(key.boxOffset, key.boxOffset + 232 * 30 * (this.generation === 6 ? 31 : 31));
    }

    scanSlots(pos1?: number, pos2?: number) {
        if (Number.isInteger(pos1)) {
            if (!Number.isInteger(pos2)) {
                pos2 = pos1;
            }
        } else {
            pos1 = 0;
            pos2 = (this.generation=== 6 ? 31 : 32) * 30;

        }

        if (this.key.isNewKey) {
            for (let i = pos1; i < pos2; ++i) {
                SaveReaderEncrypted.getPkxRaw(this.boxes1, i, this.key);
                SaveReaderEncrypted.getPkxRaw(this.boxes2, i, this.key);
            }
        } else {
            for (let i = pos1; i < pos2; ++i) {
                SaveReaderEncrypted.getPkxRaw(this.boxes2, i, this.key);
            }
        }
    }

    getPkx(pos: number): PkBase {
        var res = SaveReaderEncrypted.getPkxRaw(this.activeSlot === 0 ? this.boxes1 : this.boxes2, pos, this.key), data = res[0], ghost = res[1];
        if (data === undefined || (data[8] | data[9]) == 0)
            return undefined;
        return PkBase.makePkm(data, this.generation, (pos / 30) | 0, pos % 30, ghost);
    }

    getAllPkx() {
        var res = [];
        var tmp;
        for (var i = 0; i < (this.generation === 6 ? 930 : 960); ++i) {
            tmp = this.getPkx(i);
            if (tmp !== undefined) {
                res.push(tmp);
            }
        }
        return res;
    }

    public static getPkxRaw(boxes: Uint8Array, pos: number, key: SaveKey): [Uint8Array, boolean] {
        // Auto updates the keystream when it dumps important data!
        var pkx: Uint8Array;

        const offset = pos * 232;

        if (util.empty(key.boxKey1, offset, 232)) {
            if (util.empty(key.boxKey2, offset, 232)) {
                // We don't have any data for this slot
                util.copy(boxes, offset, key.boxKey1, offset, 232);
                return [undefined, false];
            }

            // We have a key2 only, this is supposed to be the actual key
            pkx = PkBase.decrypt(util.xor(key.boxKey2, offset, boxes, offset, 232));
            if (PkBase.verifyChk(pkx)) {
                return [pkx, false];
            } else {
                // Something is wrong. Our data didn't decrypt properly. Apparently the key is wrong.
                util.copy(zeros, 0, key.boxKey2, offset, 232);
                return [undefined, false];
            }
        }

        var data1 = util.xor(boxes, offset, key.boxKey1, offset, 232);
        var data2 = util.xor(boxes, offset, key.boxKey2, offset, 232);
        var possibleExtraKeys = [zeros, ezeros, key.blank];
        var possibleEkx1 = possibleExtraKeys.map((e) => util.xor(data1, e));
        var possibleEkx2 = possibleExtraKeys.map((e) => util.xor(data2, e));

        function tryDecrypt(possibleEkx: Uint8Array[], sourceKey: Uint8Array, sourceOffset: number, destinationKey: Uint8Array): [Uint8Array, boolean] {
            var pkx: Uint8Array;
            for (var i = 0; i < possibleExtraKeys.length; ++i) {
                pkx = PkBase.decrypt(possibleEkx[i]);
                if (PkBase.verifyChk(pkx)) {
                    util.xor(possibleExtraKeys[i], 0, sourceKey, sourceOffset, destinationKey, offset, 232);
                    return [pkx, true];
                }
            }

            return [undefined, false];
        }

        if (util.empty(key.boxKey2, offset, 232)) {
            if (possibleEkx1.some(util.empty)) {
                // Slot hasn't changed.
                return [undefined, false];
            }

            return tryDecrypt(possibleEkx1, boxes, offset, key.boxKey2);
        }

        if (possibleEkx1.some(util.empty)) {
            // Save  is the same as key 1
            return tryDecrypt(possibleEkx2, key.boxKey2, offset, key.boxKey2);
        }

        if (possibleEkx2.some(util.empty)) {
            // Save is the same as key 2
            return tryDecrypt(possibleEkx1, key.boxKey1, offset, key.boxKey1);
        }

        // Data has been observed to change twice! We can potentially get our exact keystream now!
        // Either key1 or key2 or save would be empty.
        for (let possibleEkx of possibleEkx1) {
            pkx = PkBase.decrypt(possibleEkx);
            if (!PkBase.verifyChk(pkx)) {
                continue;
            }

            if (possibleEkx2.some((e) => PkBase.verifyChk(PkBase.decrypt(e)))) {
                // Save file is empty
                util.xor(possibleEkx, 0, key.boxKey1, offset, key.boxKey2, offset, 232);
            } else {
                // Key 1 is empty
                util.xor(possibleEkx, 0, boxes, offset, key.boxKey2, offset, 232);
            }

            util.copy(zeros, 0, key.boxKey1, offset, 232);
            return [pkx, false];
        }

        for (let possibleEkx of possibleEkx2) {
            pkx = PkBase.decrypt(possibleEkx);
            if (!PkBase.verifyChk(pkx)) {
                continue;
            }

            // Key 2 is empty
            util.xor(possibleEkx, 0, boxes, offset, key.boxKey2, offset, 232);
            util.copy(zeros, 0, key.boxKey1, offset, 232);
            return [pkx, false];
        }

        // None of the three files are empty
        return [undefined, false];
    }
}
